public/pages/AnomalyCharts/containers/AnomalyHeatmapChart.tsx

```diff
@@ -0,0 +1,519 @@
+              heatmapData[0].x,
+              selectedCellIndices[1]
+            );
+          props.onHeatmapCellSelected({
+            dateRange: {
+              startDate: selectedStartDate,
+              endDate: selectedEndDate,
+            },
+            entityValue: selectedEntity,
+          } as HeatmapCell);
+        }
+      }
+    };
+
+    const getHeatmapCellDateRangeInterval = (
+      dates: string[],
+      index: number
+    ) => {
+      if (dates.length < 2) {
+        // if only less than 2 dates in X axis, it means the props.dateRange is too small.
+        // we can just use props.dateRange interval for heatmap cell
+        return props.dateRange.endDate - props.dateRange.startDate;
+      }
+      let prevIndex = index;
+      let nextIndex = index + 1;
+      if (nextIndex >= dates.length) {
+        nextIndex = index;
+        prevIndex = index - 1;
+      }
+      return (
+        moment(dates[nextIndex], HEATMAP_X_AXIS_DATE_FORMAT).valueOf() -
+        moment(dates[prevIndex], HEATMAP_X_AXIS_DATE_FORMAT).valueOf()
+      );
+    };
+
+    const getColorPaletteFlexItem = (hexCode: string) => {
+      return (
+        <EuiFlexItem grow={false} style={{ margin: '0px' }}>
+          <span
+            style={{ backgroundColor: hexCode, width: '35px', height: '8px' }}
+          />
+        </EuiFlexItem>
+      );
+    };
+
+    const handleViewEntityOptionsChange = (selectedViewOptions: any[]) => {
+      props.onHeatmapCellSelected(undefined);
+      if (isEmpty(selectedViewOptions)) {
+        // when `clear` is hit for combo box
+        setCurrentViewOptions([COMBINED_OPTIONS.options[0]]);
+        const displayTopEntityNum = get(COMBINED_OPTIONS.options[0], 'value');
+
+        const updateHeatmapPlotData = getAnomaliesHeatmapData(
+          props.anomalies,
+          props.dateRange,
+          sortByFeildValue,
+          displayTopEntityNum
+        );
+        setOriginalHeatmapData(updateHeatmapPlotData);
+        setHeatmapData(updateHeatmapPlotData);
+        setNumEntities(updateHeatmapPlotData[0].y.length);
+        setEntityViewOptions(getViewEntityOptions(updateHeatmapPlotData));
+        return;
+      }
+      const nonCombinedOptions = [] as any[];
+      for (let option of selectedViewOptions) {
+        if (currentViewOptions.includes(option)) {
+          if (!isCombinedViewEntityOption(option)) {
+            nonCombinedOptions.push(option);
+          }
+          continue;
+        }
+        if (isCombinedViewEntityOption(option)) {
+          // only allow 1 combined option
+          setCurrentViewOptions([option]);
+          const displayTopEntityNum = get(option, 'value');
+          const updateHeatmapPlotData = getAnomaliesHeatmapData(
+            props.anomalies,
+            props.dateRange,
+            sortByFeildValue,
+            displayTopEntityNum
+          );
+          setOriginalHeatmapData(updateHeatmapPlotData);
+          setHeatmapData(updateHeatmapPlotData);
```

Comment:
Could you explain why we need both `originalHeatmapData` and `heatmapData`?

Reply:
`originalHeatmapData` has data for each combined option, and `heatmapData` is only for data rendered on Heatmap. For example, if `top 20` is selected, both `originalHeatmapData` and `heatmapData` has same top 20 entities data. Then user selects 1 individual entity A, then `heatmapData` only has data for entity A, while `originalHeatmapData` still has top 20 entities data. We need `originalHeatmapData` to get individual entity options.

Reply:
i see, makes sense. Thanks